File: Assets/Colosseum/Scrpits/Skills/SkillFactory.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class SkillFactory : MonoBehaviour
{
    private static SkillFactory instance = null;

    public static SkillFactory Instance
    {
        get
        {
            if (instance == null)
            {
                instance = FindObjectOfType<SkillFactory>();
                return instance;
            }
            else
                return instance;
        }
    }
    public SkillCreator skillCreator;
    public enum SkillName { Fireball, Lightning, Frost }
    public SkillList skills;
    private void Awake()
    {
        if (instance == null)
        {
            instance = this;
        }
        skills = Instantiate(skills);
    }
    public GameObject GetSkillObject(SkillName skillName)
    {
        return skillCreator.GenerateObject((int)skillName);
    }
    public void RecycleObject(Skill skill)
    {
        skillCreator.Destroy(skill);
    }
    public SkillData GetSkillData(SkillName skillName)
    {
        foreach (SkillData data in skills.skillList)
        {
            if (data.skillName == skillName)
            {
                return data;
            }
        }
        return null;
    }
    public SkillData GetSkillData(string skillName)
    {
        foreach (SkillData data in skills.skillList)
        {
            if (data.skillName.ToString() == skillName)
            {
                return data;
            }
        }
        return null;
    }
}

File: Server/static/skill-factory.js
var Fireball = require('./skill/fireball');
var Lightning = require('./skill/lightning');
var Frost = require('./skill/frost');

module.exports = class SkillFactory{

    getSkill(data){
        if(data.skillName == 'Fireball'){
            return new Fireball(data.id,data.owner,data.skillName,data.position,data.direction);
        }else if (data.skillName == 'Lightning'){
            return new Lightning(data.id,data.owner,data.skillName,data.position,data.direction);
        }else if(data.skillName == 'Frost'){
            return new Frost(data.id,data.owner,data.skillName,data.position,data.direction);
        }
    }
    getSkillDamage(skillName){
        if(skillName == 'Fireball'){
            return 10;
        }else if (skillName == 'Lightning'){
            return 15;
        }else if(skillName == 'Frost'){
            return 15;
        }
    }

}